File: prettier-plugin-csharpier/src/Printer/Types/IfStatement.ts
import { Doc } from "prettier";
import { PrintMethod } from "../PrintMethod";
import { SyntaxToken, printSyntaxToken, SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";
import { AttributeListNode } from "./AttributeList";
import { ElseClauseNode } from "./ElseClause";

export interface IfStatementNode extends SyntaxTreeNode<"IfStatement"> {
    attributeLists: AttributeListNode[];
    ifKeyword?: SyntaxToken;
    openParenToken?: SyntaxToken;
    condition?: SyntaxTreeNode;
    closeParenToken?: SyntaxToken;
    statement?: SyntaxTreeNode;
    else?: ElseClauseNode;
}

export const printIfStatement: PrintMethod<IfStatementNode> = (path, options, print) => {
    const node = path.getValue();
    const parts: Doc[] = [printSyntaxToken(node.ifKeyword), " ", "(", path.call(print, "condition"), ")"];

    let statement = path.call(print, "statement");
    if (node.statement?.nodeType === "Block") {
        parts.push(statement);
    } else {
        parts.push(indent(concat([hardline, statement])));
    }

    if (node.else) {
        parts.push(hardline, path.call(print, "else"));
    }

    return concat(parts);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/IfStatement.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class IfStatement
{
    public static Doc Print(IfStatementSyntax node, FormattingContext context)
    {
        var docs = new List<Doc>();
        if (node.Parent is not ElseClauseSyntax)
        {
            docs.Add(ExtraNewLines.Print(node));
        }

        docs.Add(
            Token.Print(node.IfKeyword, context),
            " ",
            Doc.Group(
                Token.Print(node.OpenParenToken, context),
                Doc.Indent(
                    Doc.IfBreak(Doc.SoftLine, Doc.Null),
                    Node.Print(node.Condition, context)
                ),
                Doc.IfBreak(Doc.SoftLine, Doc.Null)
            ),
            Token.Print(node.CloseParenToken, context),
            OptionalBraces.Print(node.Statement, context)
        );

        if (node.Else != null)
        {
            docs.Add(context.NewLineBeforeElse ? Doc.HardLine : " ", Node.Print(node.Else, context));
        }

        return Doc.Concat(docs);
    }
}
